Fix 3 regular expression errors and warnings

affimnem opened the pull request:
BASE64_VALUE have a range going from 2 to 0, making the interpreter complain, and the regexp failing to match. The 'rf' formatting is also hard to get right, so I simplified by splitting the part that needs variable substitution ('f' prefix) from the raw regexp part ('r' prefix), reducing the amount of curly braces used. Finally removed some white-space that I had trouble with getting matching to work. 

TOKEN_CHAR and UNRESERVED got "SyntaxWarning: invalid escape sequence '\-'" from interpreter, so I split out a raw part from the 'f' prefixed string such that 'f' meta-characters would not interfere. 

I see now that the UNRESERVED instance is due to a previous pull request of mine. Sorry about that!

affimnem committed:
Fix 3 regular expression errors and warnings

BASE64_VALUE have a range going from 2 to 0, making the interpreter complain, and the regexp failing to match. The 'rf' formatting is also hard to get right, so I simplified by splitting the part that needs variable substitution ('f' prefix) from the raw regexp part ('r' prefix), reducing the amount of curly braces used. Finally removed some white-space that I had trouble with getting matching to work.
TOKEN_CHAR and UNRESERVED got "SyntaxWarning: invalid escape sequence '\-'" from interpreter, so I split out a raw part from the 'f' prefixed string such that 'f' meta-characters would not interfere.

NiklasBeierl commented:
> I see now that the UNRESERVED instance is due to a previous pull request of mine. Sorry about that!

Well I guess I should have been more careful in my review. :innocent: 

I guess your findings don't show up out of the blue? Can you provide examples that made you encounter these problems? I would like to use them to write tests. You can also write the tests if you want to, I just don't want to grab a mile from you after you offered an inch.  :)

affimnem commented:
I found the `BASE64_VALUE` issue due to a `script-src` failing with a valid hash value, e.g. `'sha256-4aNhtywRRWLmuerOtl4u8ZYI8BIrTA81QvsP/MhecF0='`. In my source code I first do the parsing, and I don't pick up errors from there, but go through the created policy, and flags `Unrecognized...` items. The structure is like this:
```   policy: Policy = policy_from_string(header_value)
    for directive in policy:
        if (directive.__class__.__name__ == UnrecognizedDirective.__name__):
            ...
        for val in directive.values:
            if (val.__class__.__name__ == UnrecognizedValueItem.__name__):`
               ...
```
It was the last one that hit this time. I don't know if that helps you with tests. I can look into it, but I have to postpone that a bit.
The two others I get the `SyntaxWarning` on the first time I run my application after I've modified `patterns.py` (And I did get an error on `BASE64_VALUE` as well, but I don't remember when - I debugged a bit with pdb, and did some manual testing as well.) - I'm not sure why I don't get it every time, but I'm not too familiar with Python; I just try to do enough to do what I need - and using a library is a lot easier than doing things from scratch.

affimnem committed:
Added tests for valid and invalid hash values

The valid value failed due to illegal regexp. The new tests is to make sure regressions don't occur, and that the fix works.

affimnem commented:
I've added tests for valid and invalid hash. Mainly to show that the old behaviour didn't work, and that the fixed do, for the valid hash example. I added two tests for invalid values as well, to make sure that the hash algorithm doesn't approve any algorithm/hash. Not sure that the tests follows your thinking, and I had to use a parse method to trigger the behaviour I needed to run the assert I wanted. (There are still some things I don't understand in the module code, and I'm not a Python developer and had to dig a bit to understand how to use the unit testing framework and your code in combination to verify that the (in)correct classes were instantiated.)
When it comes to the syntax warnings, I haven't made tests as I don't know if there are any that would catch what the warning complain about. Getting rid of warnings I consider a best practice (after being bit by not doing it years ago), and I see that the lib_tests_coverage.sh script gives warnings for every run, so that could be used to pick up warnings in any package file.

affimnem committed:
Fixed bad variable names

Fixed 2 instances of cut-n-paste issue.
